### ISSUE TITLE
When using instance principal `401` response from any OCI services leads to infinite loop

### ISSUE THREAD

**Issue by ashokkkannan** [oracle/oci-ruby-sdk#47]
When using instance principal, a `401` response from any OCI services leads to an infinite loop. It will come out of the loop only if it gets a non 401 response.

The problem seems to because of this function https://github.com/oracle/oci-ruby-sdk/blob/master/lib/oci/api_client.rb#L632

```
    def instance_principals_signer_wrapped_call
      max_attempts = 2

      max_attempts.times do |attempt|
        begin
          return yield
        rescue OCI::Errors::ServiceError => e
          raise if attempt >= (max_attempts - 1) # .times is zero-based
          raise if e.status_code != 401

          @signer.refresh_security_token
          retry
        end
      end
```

Sample code to show  use of `times` + `rescue-retry` leads to an infinite loop
```
max_attempts = 2
max_attempts.times do |attempt|
  begin
    puts "attempts #{attempt}"
    raise 'An error has occured'
  rescue
    raise if attempt >= (max_attempts - 1)
    retry
  end
end
```

Output:
```
 ~>ruby test.rb
attempts 0
attempts 0
attempts 0
attempts 0
attempts 0
<goes on infinitely>
```

**Comment by ashokkkannan** [comment]
Also, it may be worth adding a debug message before retrying.

**Comment by jyotisaini** [comment]
Thanks @ashokkkannan  for reporting this. We will work on this.

**Comment by jyotisaini** [comment]
Hi @ashokkkannan , this issue has been fixed with the latest release (v2.8.0). Closing this issue.

**Comment by ashokkkannan** [comment]
Thanks @jasonyin !
